Fix overview screen type error when the account has no deposit scheduled

Jinksi opened the pull request:
Fixes #5863

#### Changes proposed in this Pull Request

This PR fixes the error `TypeError` that occurs on the Payments → Overview screen when an account has no scheduled upcoming deposit.

The `nextScheduled`, `lastPaid` and `instant` properties of the type `AccountOverview.Overview`  have been changed to optional, since they can be returned as `undefined` in some cases.

https://github.com/Automattic/woocommerce-payments/blob/279046a35872335d77bea203c9b5b3085aa198a5/client/types/account-overview.d.ts#L43-L50

I've also made the `overview` prop of the `getNextDeposit()` function optional, which can be `undefined` if `useAllDepositsOverviews` has not resolved the data yet. This was throwing another error once the error above was fixed.

https://github.com/Automattic/woocommerce-payments/blob/279046a35872335d77bea203c9b5b3085aa198a5/client/components/deposits-overview/utils.ts#L21-L23

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Onboard to WCPay with a new account.
2. Enable the `_wcpay_feature_simplify_deposits_ui` feature flag: Run `update_option( '_wcpay_feature_simplify_deposits_ui', '1' );` in [WP Console](https://wordpress.org/plugins/wp-console/).
3. View the Payments → Overview screen.
4. Ensure no error is thrown in the console.
5. Ensure that the currency shown in the estimated deposits table matches the selected/default currency.

<img width="716" alt="image" src="https://user-images.githubusercontent.com/3147296/227843326-b70440ad-586c-47ed-bcc8-01cbf72f5787.png">


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Correct `AccountOverview.Overview` optional properties

`lastPaid`, `nextScheduled` and `instant` are each `undefined` in some scenarios

Jinksi committed:
Handle undefined `overview` for `getNextDeposit`

which can be `undefined` if `useAllDepositsOverviews()` `isLoading` is `true`

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5864` or branch name `fix/5863-account-overview-nextScheduled-typeerror` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/5863-account-overview-nextScheduled-typeerror) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: aaa7a2b1716e4d94f0847b5a1b988c29ffa4e43b
- Build time: 2023-03-30 20:19:56 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Add dummy changelog comment

Jinksi committed:
Add test for new accounts with undefined overview data

Jinksi committed:
Merge branch 'develop' into fix/5863-account-overview-nextScheduled-typeerror

Jinksi committed:
Merge branch 'develop' into fix/5863-account-overview-nextScheduled-typeerror

shendy-a8c reviewed (commented):
Tested with `_wcpay_feature_simplify_deposits_ui` on and off. I do see the TypeError problem on base branch and I don't see it in this PR branch, so definitely this PR fixes the issue.

I have some questions though.

shendy-a8c commented on client/types/account-overview.d.ts:
```diff
@@ -42,11 +42,11 @@ declare namespace AccountOverview {
 		currency: string;
-		lastPaid: Deposit;
-		nextScheduled: Deposit;
+		lastPaid?: Deposit;
```
Just curious, what's the difference between adding the `?` to make that property optional and adding ` | undefined`?

```suggestion
		lastPaid: Deposit | undefined;
```

Jinksi replied:
`lastPaid: Deposit | undefined;` would mean the object needs the `lastPaid` property to be present, but it can have an `undefined` value.

If we're confident that the `AccountOverview.Overview` response will _always_ have a `lastPaid` property in the response it would be more accurate to use `lastPaid: Deposit | undefined;`.

shendy-a8c replied:
Thank you for the explanation.

Jinksi replied:
Replied in https://github.com/Automattic/woocommerce-payments/pull/5864#discussion_r1152627256

shendy-a8c commented on client/types/account-overview.d.ts:
```diff
@@ -42,11 +42,11 @@ declare namespace AccountOverview {
-		nextScheduled: Deposit;
+		lastPaid?: Deposit;
+		nextScheduled?: Deposit;
 		pending: Balance;
```
I still don't quite understand fully why you make lastPaid, nextScheduled, and instant optional and why only those 3? Why didn't you make `pending` and `available` also optional?

Looking at https://github.com/Automattic/woocommerce-payments/blob/fix/5863-account-overview-nextScheduled-typeerror/client/data/deposits/selectors.js#L86-L93
potentially all of them can be optional.

However, the next line https://github.com/Automattic/woocommerce-payments/blob/fix/5863-account-overview-nextScheduled-typeerror/client/data/deposits/selectors.js#L97
is setting each key: `lastPaid`, `nextScheduled`, `pending`, `available`, and `instant`.

Also, looking at server code, the value of `deposit.next_scheduled` will always be defined. At most, it'll be empty array, so I don't understand why we have to handle when it's undefined.

shendy-a8c replied:
I am just rubber ducking. Now I can see why. It's because when it's an empty array, it won't this [forEach](https://github.com/Automattic/woocommerce-payments/blob/fix/5863-account-overview-nextScheduled-typeerror/client/data/deposits/selectors.js#L82) won't even run even once, right?

shendy-a8c replied:
I also can now see why `pending` and `available` are always defined.

Jinksi replied:
As you mentioned, this is where the `undefined` value originates from: `client/data/deposits/selectors.js` `getAllDepositsOverviews()`:

https://github.com/Automattic/woocommerce-payments/blob/b644b98c5c3a83c0969749762d64f7d25c18bda3/client/data/deposits/selectors.js#L86-L93

So I think you're correct that we can safely use `lastPaid: Deposit | undefined;`.

We can also update `pending`, `available` and `instant`.

> Also, looking at server code, the value of deposit.next_scheduled will always be defined.

This `AccountOverview.Overview` type is being used by `AccountOverview.OverviewsResponse.overview.currencies`:

https://github.com/Automattic/woocommerce-payments/blob/b644b98c5c3a83c0969749762d64f7d25c18bda3/client/types/account-overview.d.ts#L52-L58

So the response from the server is being restructured in the `getAllDepositsOverviews()` code above, and the `undefined` value for these properties is being introduced there.

Logging the value of `currencies` from [this line](https://github.com/Automattic/woocommerce-payments/blob/b644b98c5c3a83c0969749762d64f7d25c18bda3/client/data/deposits/selectors.js#L101) confirms that `nextScheduled` can be `undefined` but is present on the currency object:

<img width="831" alt="image" src="https://user-images.githubusercontent.com/3147296/228704329-de2d6b01-40c3-45d6-bbad-1f096351eef5.png">

I'll update the `AccountOverview.Overview` type to ensure all properties can be `undefined`. ~Since this relates only to the `currencies` array, perhaps we could extend `AccountOverview.Overview` instead?~ This type is only used by `AccountOverview.OverviewsResponse.overviews.currencies` AFAICT.

Jinksi replied:
I've updated this in cebf521.

@shendy what do you think of this solution?

shendy-a8c replied:
Looks good @Jinksi.

shendy-a8c reviewed (approved):
Looks good!

Jinksi committed:
Let all AccountOverview.Overview values be `undefined`

but ensure the property is present on the Overview object.

the logic defining these as `undefined` is found in client/data/deposits/selectors.js#L86-L93

Jinksi committed:
Merge branch 'develop' into fix/5863-account-overview-nextScheduled-typeerror

Jinksi committed:
Merge branch 'develop' into fix/5863-account-overview-nextScheduled-typeerror

Jinksi committed:
Merge branch 'develop' into fix/5863-account-overview-nextScheduled-typeerror
